### DISCUSSION TITLE
How to handle GraphQL immutable classes?

### DISCUSSION THREAD

**Discussion by arjunballa** [graphql-java/graphql-java#2533] [Q&A]
I have a requirement to create GraphQL Schema programmatically and I get the information needed to create different objects in a piecemeal manner.

I am trying to add a list of `TypeDefinition` objects to `TypeDefinitionRegistry` and later convert that into `GraphQLSchema` using the below code.

```
GraphQLSchema graphQLSchema = UnExecutableSchemaGenerator.makeUnExecutableSchema(typeDefinitionRegistry)
```

Initially I have only name to create `ObjectTypeDefintion`, `FieldDefinition`, `InputValueDefinition`, `Type` objects. 

I understand GraphQLXXX classes like `GraphQLSchema`, `GraphQLObjectType`, `GraphQLFieldDefinition`, `GraphQLArgument`, `GraphQLOutputType` are more restrictive but I was expecting `ObjectTypeDefintion`, `FieldDefinition`, `InputValueDefinition`, `Type ` classes to be a bit relaxed with the builder method to change the original state.

May I know the reason for making classes immutable?

Is it to make them thread-safe and support functional programming? 

I think there should be a parallel set of mutable classes and leave it to the developer to use at his own risk and make them thread-safe.

As the classes are immutable now I am maintaining the internally and then copying it to  TypeDefinitionRegistry once I have all the info which is unnecessary and I am not taking complete advantage of the library.

If the classes are internal and you know that all use cases can work with immutable objects then it makes sense to make them immutable but for public classes providing only immutable versions does not make much sense.

**Comment by bbakerman** [comment]
The class ObjectTypeDefintion are AST nodes that come from parsing a text document - they are not types - they are representations of grammar text.  You dont create new schemas programtically via them.

Honestly if you use these class, you would be better to "print a string, parse it and haver them auto created"

The way you programmatically add types to a schema is via the `GraphqlXXXType` classes and `graphql.schema.SchemaTransformer`

See https://www.graphql-java.com/documentation/v17/schema/ and the section "Changing Schema"

**Comment by arjunballa** [comment]
https://www.graphql-java.com/documentation/v17/schema/ and the section "Changing Schema" does not work in my case.

I don't have complete information to construct GraphqlXXXType objects to start with. I get that in piecemeal manner.

For example to construct GraphQLObjectType all I have initially is `name` and I need to construct GraphQLObjectType with just `name ` and later fill it up with other info but framework dos not allow it because it does below validations.

```
 assertValidName(name);
assertNotNull(fieldDefinitions, () -> "fieldDefinitions can't be null");
assertNotNull(interfaces, () -> "interfaces can't be null");
```

**Reply by bbakerman** [reply]
Then hold onto the "builder" objects until you do have all the required information.
